lib/statsd/instrument/batched_udp_sink.rb

```diff
@@ -61,111 +56,85 @@ def shutdown(*args)
-              wakeup_thread
-            end
-
-            # A SizedQueue would be perfect, except that it doesn't have a timeout
-            # Ref: https://bugs.ruby-lang.org/issues/18774
-            if @buffer.size >= @buffer_capacity
-              StatsD.logger.warn do
-                "[#{self.class.name}] Max buffer size reached (#{@buffer_capacity}), pausing " \
-                  "thread##{Thread.current.object_id}"
-              end
-              before = Process.clock_gettime(Process::CLOCK_MONOTONIC, :float_millisecond)
-              @monitor.synchronize do
-                while @buffer.size >= @buffer_capacity && @dispatcher_thread.alive?
-                  @condition.wait(0.01)
-                end
-              end
-              duration = Process.clock_gettime(Process::CLOCK_MONOTONIC, :float_millisecond) - before
-              StatsD.logger.warn do
-                "[#{self.class.name}] thread##{Thread.current.object_id} resumed after #{duration.round(2)}ms"
-              end
-            end
-          else
-            flush
+          if !thread_healthcheck || !@buffer.push_nonblock(datagram)
+            # The buffer is full or the thread can't be respaned,
+            # we'll send the datagram synchronously
+            @udp_sink << datagram
           end
 
           self
         end
 
-        def shutdown(wait = @flush_interval * 2)
+        def shutdown(wait = 2)
           @interrupted = true
+          @buffer.close
           if @dispatcher_thread&.alive?
             @dispatcher_thread.join(wait)
-          else
-            flush
           end
+          flush(blocking: false)
         end
 
         private
 
-        def wakeup_thread
-          begin
-            @monitor.synchronize do
-              @condition.signal
-            end
-          rescue ThreadError
-            # Can't synchronize from trap context
-            Thread.new { wakeup_thread }.join
-            return
-          end
-
-          begin
-            @dispatcher_thread&.run
-          rescue ThreadError # Somehow the thread just died
-            thread_healthcheck
-          end
-        end
-
         NEWLINE = "\n".b.freeze
-        def flush
-          return if @buffer.empty?
 
-          datagrams = @buffer.shift(@buffer.size)
-
-          until datagrams.empty?
-            packet = String.new(datagrams.shift, encoding: Encoding::BINARY, capacity: @max_packet_size)
+        def flush(blocking:)
+          packet = "".b
+          next_datagram = nil
+          until @buffer.closed? && @buffer.empty? && next_datagram.nil?
+            next_datagram ||= blocking ? @buffer.pop : @buffer.pop_nonblock
+            if next_datagram.nil?
+              # Either we are in non blocking mode and there was no datagram in the buffer,
+              # or either we were in blocking mode and the queue was closed.
+              # In both case there is nothing left to do.
+              break
+            end
```

Comment:
I'd split this up since these are two pretty different reasons:
```suggestion
            if blocking
              next_datagram ||= @buffer.pop
              break if next_datagram.nil? # queue was closed
            else
              next_datagram ||= @buffer.pop_nonblock
              break if next_datagram.nil? # no datagram in buffer
            end
          end
```